tools/codegen/gen.py

```diff
@@ -276,10 +276,21 @@ def func(f: NativeFunction) -> Optional[str]:
                     cuda_guard = """\
     // DeviceGuard omitted
 """
+            mem_overlap_check = ""
+            if (dispatch == "CPU" or dispatch == "CUDA") and ('backward' not in impl_name):
+                if f.func.name.name.inplace:
+                    mem_overlap_check = """\
+    at::assert_no_internal_overlap(self);
```

Comment:
For certain in-place operations, we should allow inputs that do have internal memory overlap. This is because those in-place operations are valid and to preserve BC. For example:
```
x = torch.randn(3).expand(3, 3)
x.zero_()
```
I'm not sure if we should just make a decision to break BC for all of these cases but am open to discussion.